### PR TITLE
Add layout type in Lambda

### PR DESCRIPTION
**Pull request by Ekdohibs**
For now, we simply have the type `type layout = Pvalue of value_kind`, to prepare for the addition of layouts everywhere. Besides, this also adds several pre-defined layouts (often equal to `Pvalue Pgenval`), so that we know why `Pgenval` was used, for instance whether it was a true `top` value or whether it was the kind of a function.

There are quite a number of `layout_top` instances, which will need to be removed with further PRs, as there is no `top` layout once we start getting non-value layouts. Hopefully, this PR is the one with the most global effect, and other PRs can start cleaning the code with a smaller scope at a time.

I'd especially like reviews concerning the `translmod`, `translobj` and `translclass` files, where there were a lot of instances of `Pgenval`, and I was unsure how to categorize a lot of them (so they ended up as `layout_top`).

### CODE REVIEW

**Comment by mshinwell** [comment]
@lthls is going to review this.

File: ocaml/lambda/translobj.ml [hunk 1]
```diff
@@ -95,7 +95,7 @@ let transl_label_init_general f =
          let const =
            Lprim (Popaque, [Lconst c], Debuginfo.Scoped_location.Loc_unknown)
          in
-         Llet(Alias, Pgenval, id, const, expr))
+         Llet(Alias, Lambda.structured_constant_layout c, id, const, expr))
```
Comment (lthls) [inline ocaml/lambda/translobj.ml:98]: We should put a CR mentioning that we must not have a layout too precise (that's why we added the `Popaque` wrapping). Currently this should end up as `Pgenval` so it should be fine.

File: ocaml/lambda/translobj.ml [hunk 2]
```diff
@@ -186,7 +186,7 @@ let oo_wrap env req f x =
                         [lambda_unit; lambda_unit; lambda_unit],
                         Loc_unknown)
                 in
-                Llet(StrictOpt, Pgenval, id,
+                Llet(StrictOpt, Lambda.layout_class, id,
```
Comment (lthls) [inline ocaml/lambda/translobj.ml:189]: I don't think this is actually a class. No idea what it is though.

**Review by lthls — CHANGES_REQUESTED** [review]
I'm made some comments. I got a bit tired of commenting every single `layout_top`, and most of them should be fixed in #1084 anyway.

File: ocaml/lambda/translobj.ml (hunk 2 above)
Comment (Ekdohibs) [reply ocaml/lambda/translobj.ml:192]:
```suggestion
                Llet(StrictOpt, Lambda.layout_block, id,
```

File: native_toplevel/opttoploop.ml [hunk 3]
```diff
@@ -122,7 +122,7 @@ let close_phrase lam =
              [Lprim (Pgetglobal glb, [], Loc_unknown)],
              Loc_unknown)
     in
-    Llet(Strict, Pgenval, id, glob, l)
+    Llet(Strict, Lambda.layout_top, id, glob, l)
```
[the remark this answers is not in the document]
Comment (Ekdohibs) [reply native_toplevel/opttoploop.ml:125]:
```suggestion
    Llet(Strict, Lambda.layout_module_field, id, glob, l)
```

**Review by lthls — APPROVED** [review]

**Comment by Ekdohibs** [comment]
Rebased on current main.

**Comment by ccasin** [comment]
I took a quick look at the changes under `ocaml`, and they look reasonable to me in terms of compatibility with the approach we're taking in the type systems side of this.  Some high-level thoughts:

- This PR adds a `layout` datatype in `Lambda.mli`.  I think the idea here is that this will mirror the "sorts" in our layout design, but adding more information in the value case.  (Sorts are, roughly, the layouts that capture just enough information to know how to compile a type).  This seems reasonable.

  In particular, Lambda's view of "value" and its sublayouts, captured by the existing `value_kind` type, is already more fine-grained that the layouts in the typechecker.  An alternative design would be to use the type system's sorts and pair them with additional information in the value case, but the design here seems cleaner.

- A minor difference from what is done in the [current type system prototype](https://github.com/ccasin/ocaml-jst/tree/unboxed-clean-rebase) is that this PR changes `Typeopt.value_kind` into `Typeopt.layout`.  I think I slightly prefer letting `value_kind` be just `value_kind` and adding a second function which dispatches to it in the value case, mostly because `value_kind` is already quite complicated.  Though admittedly that approach does require some care to avoid recursively calling `value_kind` on non-value types, so I don't feel strongly.

- This PR adds changes the `value_kinds` in the lambda datatype to `layouts`.  This seems right (though in most of these places we'll allow only values in the short term).  Obviously there will be some work to do to propagate layouts to the right places when we merge this with the front end bits and add more layouts, but I don't see any fundamental obstacles - I already did a lot of the needed work to deal with the (rather silly) way we are compiling `void` in the current type system prototype.

**Comment by ccasin** [comment]
I realized after writing this that these two points are related:

> I think I slightly prefer letting `value_kind` be just `value_kind` and adding a second function which dispatches to it in the value case, mostly because `value_kind` is already quite complicated.
> ...
> In particular, we have taken some care to store layout information in the typed tree to avoid the need for the translation to lambda to need to recompute layouts of types

Because our goal is to avoid ever needing to calculate the layout for a `type_expr` during the translation to lambda by storing enough information in the typed tree, we should have two functions with different types:

```
val layout : Env.t -> Types.sort -> Types.type_expr -> Lambda.layout

val value_kind  : Env.t -> Types.type_expr -> Lambda.value_kind
```

The idea being that the former doesn't actually look at the `type_expr`, it just passes it to `value_kind` in the case where the sort is value.

Does that make sense?

**Comment by Ekdohibs** [comment]
Ah yes, I see, this would probably be better indeed as it would avoid having to look at the `type_expr`. This is almost what is currently being done, since the `layout` function currently just defers to the `value_kind` function, adding a `Pvalue` wrapper around it (but `value_kind` is no longer exposed).

There might be a few cases where we need some more changes to the typed tree to avoid having to recompute the layouts in all cases: these correspond to all uses of `function_return_layout` in the PR, where the layout will have to be stored in the typed tree explicitly.

**Review by mshinwell — APPROVED** [review]

**Pull request merged**